### ISSUE TITLE
Tokenization Inconsistency Between Training and Inference (an example with Open-Orca/Mistral-7B-OpenOrca)

### ISSUE THREAD

**Issue by tokestermw** [axolotl-ai-cloud/axolotl#834]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Training on `axolotl` and running on inference on HF gives inconsistent tokenization of the prompts.

For example, take this chat.

```python
chat = [
    {
        "role": "system",
        "content": "You are MistralOrca, a large language model trained by Alignment Lab AI. Write out your reasoning step-by-step to be sure you get the right answers!",
    },
    {"role": "user", "content": "How are you?"},
    {"role": "assistant", "content": "I am doing well!"},
]
```

Add the tokenizer

```python
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained(
    "Open-Orca/Mistral-7B-OpenOrca",
    use_fast=False,  # or True
)
```

Get the default prompt format if `apply_chat_template` exists

```python
apply_chat_template_prompt = tokenizer.apply_chat_template(
    chat,
    tokenize=False,
    add_generation_prompt=False,
)
```

or if we're using FastChat aka ("ShareGPTPrompter")

```python
from fastchat.conversation import get_conv_template

conv_template = get_conv_template("mistral-7b-openorca")
role_id = 0
for message in chat + (
    [{"role": "assistant", "content": ""}] if add_generation_prompt else []
):
    if message["role"] == "system":
        conv_template.set_system_message(message["content"])
    else:
        conv_template.append_message(
            conv_template.roles[role_id], message["content"]
        )
        role_id = 0 if role_id else 1  # switch back and forth
fastchat_template_prompt = conv_template.get_prompt()
```

These two prompts are the same, hence, tokens will be the same.

```python
assert fastchat_template_prompt == apply_chat_template_prompt
```

When we build the prompt through `axolotl`, the last two assertions should pass (especially because we are still using the same `mistral-7b-openorca` template from fastchat).

```python
from axolotl.cli import load_cfg
from axolotl.prompt_strategies import load
from axolotl.monkeypatch.fastchat_conversation_turns import (
    add_get_turns_to_conversation,
)

add_get_turns_to_conversation()

config_path = os.path.join("tests/axolotl", "assets", "config.yaml")
cfg = load_cfg(config_path)
ds_cfg = cfg["datasets"][0]
strategy = load("sharegpt", tokenizer, cfg, ds_cfg)

role_map = {"user": "human", "assistant": "gpt", "system": "system"}
mapped_chat = [{"from": role_map[i["role"]], "value": i["content"]} for i in chat]

axolotl_tokenized_prompt = strategy.tokenize_prompt({"conversations": mapped_chat})[
    "input_ids"
]
axolotl_decoded_prompt = tokenizer.decode(axolotl_tokenized_prompt)
print("*** axolotl_decoded_prompt ***")
print(axolotl_decoded_prompt)

fastchat_tokenized_prompt = tokenizer.encode(fastchat_template_prompt)
fastchat_decoded_prompt = tokenizer.decode(fastchat_tokenized_prompt)
print("*** fastchat_decoded_prompt ***")
print(fastchat_decoded_prompt)

assert fastchat_tokenized_prompt == axolotl_tokenized_prompt
assert fastchat_decoded_prompt == axolotl_decoded_prompt
```

where the dataset config is something like

```yaml
  - path: ...
    type: sharegpt
    conversation: mistral-7b-openorca
    ds_type: json
```

### Current behaviour

The prompts look as follows:

axolotl

```
<s><|im_start|> system
You are MistralOrca, a large language model trained by Alignment Lab AI. Write out your reasoning step-by-step to be sure you get the right answers!<|im_end|>
<|im_start|> user
How are you?<|im_end|>
<|im_start|> assistant
I am doing well!<|im_end|>

```

and fastchat

```
<s><|im_start|> system
You are MistralOrca, a large language model trained by Alignment Lab AI. Write out your reasoning step-by-step to be sure you get the right answers!<|im_end|><|im_start|> user
How are you?<|im_end|><|im_start|> assistant
I am doing well!<|im_end|>

```

The axolotl prompt has two extra tokens after each `<|im_end|>`: `[28705, 13]`. The new line (13) is expected, but there is a mysterious space(?) token (28705) that's introduced (GitHub gobbles it up so we can't really see it above). I think it's a quirk of the LLamaTokenizer but not sure.

In general, HF tokenizers are a bit destructive, as in, `text != tokenizer.decode(tokenizer.encode(text))`.

And axolotl tokenizes and de-tokenize _per message_ instead of per chat to build the prompts. This makes sense because it's easier to add loss masking and such. But a drawback is that the tokenization could get inconsistent.

The reason `[28705, 13]` does not appear in the fastchat example is that the `<|im_start|>` token has `lstrip` and `rstrip` as `true` so it gobbles up the spaces. When tokenizing per message, the `<|im_start|>` is not there to gobble them up.

Consequently, what I've found by training a bunch of models is that it's very difficult to stop the generations properly with `<|im_end|>`.

### Steps to reproduce

See above (I've tried to add as many details as possible :))

### Config yaml

```yaml
base_model: Open-Orca/Mistral-7B-OpenOrca
base_model_config: Open-Orca/Mistral-7B-OpenOrca
model_type: MistralForCausalLM
tokenizer_type: AutoTokenizer
tokenizer_config: Open-Orca/Mistral-7B-OpenOrca
is_mistral_derived_model: true

load_in_8bit: false
load_in_4bit: false
strict: false

datasets:
  - path: [INSERT DATASET HERE]
    type: sharegpt
    conversation: mistral-7b-openorca
    ds_type: json

dataset_prepared_path:
val_set_size: 0.01
output_dir: ./out

sequence_len: 8192
sample_packing: true
pad_to_sequence_len: true

adapter: lora
lora_model_dir:
lora_r: 32
lora_alpha: 16
lora_dropout: 0.05
lora_target_linear: false
lora_fan_in_fan_out:
lora_target_modules:
  - q_proj
  - v_proj

wandb_project:
wandb_entity:
wandb_watch:
wandb_run_id:
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 2
eval_batch_size: 2
num_epochs: 3
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 0.01

train_on_inputs: false
group_by_length: false
bf16: false
fp16: false
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
eval_steps: 20
eval_table_size: 5
eval_table_max_new_tokens: 128
eval_sample_packing: false
save_steps: 5
debug:
weight_decay: 0.0
fsdp:
fsdp_config:
```

### Possible solution

* I think this can be fixed potentially by setting lstrip and rstrip to true on `<|im_end|>` as well, but looks like that was changed, maybe it had other issues (https://huggingface.co/Open-Orca/Mistral-7B-OpenOrca/commit/17572416df27482d71dda9ea6bdea1733d8cee5d)
* In general, it might be better to add loss masks by tokenizing once per chat, then doing some string matching, but that may also be finicky
* Hacky way is to rstrip if `content.endswith('<|im_end|>\n')`; https://github.com/OpenAccess-AI-Collective/axolotl/blob/main/src/axolotl/prompt_tokenizers.py#L364 which so far have worked for me

### Which Operating Systems are you using?

- [X] Linux
- [X] macOS
- [ ] Windows

### Python Version

3.10

### axolotl branch-commit

main/6dc68a653f1b031b9c03613a399dd3c7f5c904f2

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by mhenrichsen** [comment]
Using the prompt templates should have fixed this issue.
